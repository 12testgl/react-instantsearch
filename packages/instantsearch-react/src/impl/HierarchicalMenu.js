import React, {PropTypes, Component} from 'react';
import themeable from 'react-themeable';

import {
  hierarchicalItemPropType,
  hierarchicalItemsPropType,
  selectedItemsPropType,
} from '../propTypes';
import MenuLink from './MenuLink';

//
// const defaultTranslations = {
//   showMore: extended => extended ? 'Show less' : 'Show more',
// };

const defaultTheme = {
  root: 'HierarchicalMenu',
  list: 'HierarchicalMenu__list',
  item: 'HierarchicalMenu__item',
  itemSelected: 'HierarchicalMenu__item--selected',
  itemSelectedParent: 'HierarchicalMenu__item--selectedParent',
  itemLink: 'HierarchicalMenu__item__link',
  itemValue: 'HierarchicalMenu__item__value',
  itemCount: 'HierarchicalMenu__item__count',
  showMore: 'HierarchicalMenu__showMore',
};

class HierarchicalMenuItem extends Component {
  static propTypes = {
    theme: PropTypes.object.isRequired,
    // translations: PropTypes.object,
    item: hierarchicalItemPropType.isRequired,
    selectedItems: selectedItemsPropType.isRequired,
    refine: PropTypes.func.isRequired,
    createURL: PropTypes.func.isRequired,
  };

  // We could have only one onClick method since MenuLink passes its onClick
  // handler the item that was provided as props, but having it here might make
  // more sense since we also use createURL the same way in the same component.
  onClick = item => {
    this.props.refine(item.value);
  };

  render() {
    const {
      item,
      selectedItems,
      refine,
      createURL,
      theme,
    } = this.props;

    const th = themeable(theme);

    return (
      <li
        {...th(
          'item',
          'item',
          selectedItems.indexOf(item.value) !== -1 && 'itemSelected',
          item.children && 'itemSelectedParent'
        )}
      >
        <MenuLink
          theme={{
            root: theme.itemLink,
            value: theme.itemValue,
            count: theme.itemCount,
          }}
          item={item}
          href={createURL(item.value)}
          onClick={this.onClick}
        />

        {item.children &&
          <HierarchicalMenuList
            theme={theme}
            refine={refine}
            createURL={createURL}
            items={item.children}
            selectedItems={selectedItems}
          />
        }
      </li>
    );
  }
}

class HierarchicalMenuList extends Component {
  static propTypes = {
    theme: PropTypes.object.isRequired,
    // translations: PropTypes.object,
    items: hierarchicalItemsPropType.isRequired,
    selectedItems: selectedItemsPropType.isRequired,
    refine: PropTypes.func.isRequired,
    createURL: PropTypes.func.isRequired,
  };

  render() {
    const {
      items,
      selectedItems,
      refine,
      createURL,
      theme,
    } = this.props;

    const th = themeable(theme);

    return (
      <ul {...th('list', 'list')}>
        {items.map(item =>
          <HierarchicalMenuItem
            key={item.value}
            theme={theme}
            refine={refine}
            createURL={createURL}
            item={item}
            selectedItems={selectedItems}
          />
        )}
      </ul>
    );
  }
}

class HierarchicalMenu extends Component {
  static propTypes = {
    theme: PropTypes.object,
    refine: PropTypes.func.isRequired,
    createURL: PropTypes.func.isRequired,
    items: hierarchicalItemsPropType,
    selectedItems: selectedItemsPropType,
  };

  static defaultProps = {
    theme: defaultTheme,
  };

  render() {
    const {items, selectedItems, createURL, refine, theme} = this.props;
    if (!items) {
      return null;
    }

    const th = themeable(theme);

    return (
      <div {...th('root', 'root')}>
        <HierarchicalMenuList
          theme={theme}
          items={items}
          selectedItems={selectedItems}
          refine={refine}
          createURL={createURL}
        />
      </div>
    );
  }
}

export default HierarchicalMenu;
