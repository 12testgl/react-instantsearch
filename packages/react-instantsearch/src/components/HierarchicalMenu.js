import PropTypes from 'prop-types';
import React, { Component } from 'react';
import { pick } from 'lodash';

import translatable from '../core/translatable';
import classNames from './classNames.js';
import List from './List';
import Link from './Link';
import BaseWidget from './BaseWidget';

const widgetClassName = 'HierarchicalMenu';
const cx = classNames(widgetClassName);

const itemsPropType = PropTypes.arrayOf(
  PropTypes.shape({
    label: PropTypes.string.isRequired,
    value: PropTypes.string,
    count: PropTypes.number.isRequired,
    items: (...args) => itemsPropType(...args),
  })
);

class HierarchicalMenu extends Component {
  static propTypes = {
    translate: PropTypes.func.isRequired,
    refine: PropTypes.func.isRequired,
    createURL: PropTypes.func.isRequired,
    canRefine: PropTypes.bool.isRequired,
    items: itemsPropType,
    showMore: PropTypes.bool,
    limitMin: PropTypes.number,
    limitMax: PropTypes.number,
    transformItems: PropTypes.func,
    header: PropTypes.node,
    footer: PropTypes.node,
  };

  static contextTypes = {
    canRefine: PropTypes.func,
  };

  componentWillMount() {
    if (this.context.canRefine) this.context.canRefine(this.props.canRefine);
  }

  componentWillReceiveProps(props) {
    if (this.context.canRefine) this.context.canRefine(props.canRefine);
  }

  renderItem = item => {
    const { createURL, refine } = this.props;

    return (
      <Link
        {...cx(['link'])}
        onClick={() => refine(item.value)}
        href={createURL(item.value)}
      >
        <span {...cx(['label'])}>{item.label}</span>{' '}
        <span {...cx(['count'])}>{item.count}</span>
      </Link>
    );
  };

  render() {
    const { header, footer } = this.props;

    return (
      <BaseWidget
        widgetClassName={widgetClassName}
        header={header}
        footer={footer}
      >
        <List
          renderItem={this.renderItem}
          cx={cx}
          {...pick(this.props, [
            'translate',
            'items',
            'showMore',
            'limitMin',
            'limitMax',
            'isEmpty',
            'canRefine',
          ])}
        />
      </BaseWidget>
    );
  }
}

export default translatable({
  showMore: extended => (extended ? 'Show less' : 'Show more'),
})(HierarchicalMenu);
