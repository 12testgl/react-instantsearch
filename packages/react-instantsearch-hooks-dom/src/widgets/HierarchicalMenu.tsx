import React from 'react';
import { useHierarchicalMenu } from 'react-instantsearch-hooks';

import { HierarchicalMenu as HierarchicalMenuUiComponent } from '../ui/HierarchicalMenu';
import { cx } from '../ui/lib/cx';
import { ShowMoreButton } from '../ui/ShowMoreButton';

import type { HierarchicalMenuProps as HierarchicalMenuUiComponentProps } from '../ui/HierarchicalMenu';
import type { UseHierarchicalMenuProps } from 'react-instantsearch-hooks';

export type HierarchicalMenuProps = Omit<
  HierarchicalMenuUiComponentProps,
  'items' | 'createURL' | 'hasItems' | 'onNavigate' | 'showMoreButton'
> &
  UseHierarchicalMenuProps;

export function HierarchicalMenu({
  attributes,
  limit,
  rootPath,
  separator,
  showMore,
  showMoreLimit,
  showParentLevel,
  sortBy,
  transformItems,
  ...props
}: HierarchicalMenuProps) {
  const {
    items,
    canRefine,
    canToggleShowMore,
    createURL,
    isShowingMore,
    refine,
    toggleShowMore,
  } = useHierarchicalMenu(
    {
      attributes,
      limit,
      rootPath,
      separator,
      showMore,
      showMoreLimit,
      showParentLevel,
      sortBy,
      transformItems,
    },
    {
      $$widgetType: 'ais.hierarchicalMenu',
    }
  );

  return (
    <HierarchicalMenuUiComponent
      {...props}
      items={items}
      hasItems={canRefine}
      createURL={createURL}
      onNavigate={refine}
      showMoreButton={
        showMore && (
          <ShowMoreButton
            className={cx(
              'ais-HierarchicalMenu-showMore',
              !canToggleShowMore && 'ais-HierarchicalMenu-showMore--disabled'
            )}
            disabled={!canToggleShowMore}
            onClick={toggleShowMore}
            isShowingMore={isShowingMore}
          />
        )
      }
    />
  );
}
