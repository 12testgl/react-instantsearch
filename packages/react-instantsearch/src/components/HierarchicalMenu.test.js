/* eslint-env jest, jasmine */

import React from 'react';
import renderer from 'react-test-renderer';
import Enzyme, { mount } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
Enzyme.configure({ adapter: new Adapter() });

import HierarchicalMenu from './HierarchicalMenu';

describe('HierarchicalMenu', () => {
  it('default hierarchical menu', () => {
    const tree = renderer
      .create(
        <HierarchicalMenu
          refine={() => null}
          createURL={() => '#'}
          items={[
            {
              value: 'white',
              count: 10,
              label: 'white',
              items: [
                { value: 'white1', label: 'white1', count: 3 },
                { value: 'white2', label: 'white2', count: 4 },
              ],
            },
            { value: 'black', count: 20, label: 'black' },
            { value: 'blue', count: 30, label: 'blue' },
          ]}
          limitMin={2}
          limitMax={4}
          showMore={true}
          canRefine={true}
        />
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('applies translations', () => {
    const tree = renderer
      .create(
        <HierarchicalMenu
          refine={() => null}
          createURL={() => '#'}
          items={[
            {
              value: 'white',
              count: 10,
              label: 'white',
              items: [
                { value: 'white1', label: 'white1', count: 3 },
                { value: 'white2', label: 'white2', count: 4 },
              ],
            },
            { value: 'black', count: 20, label: 'black' },
            { value: 'blue', count: 30, label: 'blue' },
          ]}
          translations={{
            showMore: ' display more',
          }}
          canRefine={true}
        />
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('refines its value on change', () => {
    const refine = jest.fn();
    const wrapper = mount(
      <HierarchicalMenu
        refine={refine}
        createURL={() => '#'}
        items={[
          {
            value: 'white',
            count: 10,
            label: 'white',
            items: [
              { value: 'white1', label: 'white1', count: 3 },
              { value: 'white2', label: 'white2', count: 4 },
            ],
          },
          { value: 'black', count: 20, label: 'black' },
          { value: 'blue', count: 30, label: 'blue' },
        ]}
        canRefine={true}
      />
    );

    const itemParent = wrapper.find('.ais-HierarchicalMenu-item--parent');

    expect(itemParent).toHaveLength(1);

    itemParent
      .find('Link')
      .first()
      .simulate('click');
    expect(refine.mock.calls).toHaveLength(1);
    expect(refine.mock.calls[0][0]).toEqual('white');

    wrapper.unmount();
  });

  it('should respect defined limits', () => {
    const refine = jest.fn();
    const wrapper = mount(
      <HierarchicalMenu
        refine={refine}
        createURL={() => '#'}
        items={[
          { value: 'white', count: 10, label: 'white' },
          { value: 'black', count: 20, label: 'black' },
          { value: 'blue', count: 30, label: 'blue' },
          { value: 'green', count: 30, label: 'green' },
          { value: 'cyan', count: 30, label: 'cyan' },
        ]}
        limitMin={2}
        limitMax={4}
        showMore={true}
        canRefine={true}
      />
    );

    const items = wrapper.find('li');

    expect(items).toHaveLength(2);

    wrapper.find('button').simulate('click');

    expect(wrapper.find('li')).toHaveLength(4);

    wrapper.unmount();
  });

  it('should disabled show more when no more item to display', () => {
    const refine = jest.fn();
    const wrapper = mount(
      <HierarchicalMenu
        refine={refine}
        createURL={() => '#'}
        items={[
          { value: 'white', count: 10, label: 'white' },
          { value: 'black', count: 20, label: 'black' },
        ]}
        limitMin={2}
        limitMax={4}
        showMore={true}
        canRefine={true}
      />
    );

    const items = wrapper.find('li');

    expect(items).toHaveLength(2);

    expect(wrapper.find('button')).toBeDefined();

    wrapper.unmount();
  });
});
