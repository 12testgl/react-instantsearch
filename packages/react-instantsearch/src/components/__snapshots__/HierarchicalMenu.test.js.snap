// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HierarchicalMenu applies translations 1`] = `
<ul
  className="list false"
>
  <li
    className="item   item--parent"
  >
    <a
      className="link"
      href="#"
      onClick={[Function]}
    >
      <span
        className="label"
      >
        white
      </span>
       
      <span
        className="count"
      >
        10
      </span>
    </a>
    <ul
      className="list"
    >
      <li
        className="item   "
      >
        <a
          className="link"
          href="#"
          onClick={[Function]}
        >
          <span
            className="label"
          >
            white1
          </span>
           
          <span
            className="count"
          >
            3
          </span>
        </a>
      </li>
      <li
        className="item   "
      >
        <a
          className="link"
          href="#"
          onClick={[Function]}
        >
          <span
            className="label"
          >
            white2
          </span>
           
          <span
            className="count"
          >
            4
          </span>
        </a>
      </li>
    </ul>
  </li>
  <li
    className="item   "
  >
    <a
      className="link"
      href="#"
      onClick={[Function]}
    >
      <span
        className="label"
      >
        black
      </span>
       
      <span
        className="count"
      >
        20
      </span>
    </a>
  </li>
  <li
    className="item   "
  >
    <a
      className="link"
      href="#"
      onClick={[Function]}
    >
      <span
        className="label"
      >
        blue
      </span>
       
      <span
        className="count"
      >
        30
      </span>
    </a>
  </li>
</ul>
`;

exports[`HierarchicalMenu default hierarchical menu 1`] = `
<ul
  className="list false"
>
  <li
    className="item   item--parent"
  >
    <a
      className="link"
      href="#"
      onClick={[Function]}
    >
      <span
        className="label"
      >
        white
      </span>
       
      <span
        className="count"
      >
        10
      </span>
    </a>
    <ul
      className="list"
    >
      <li
        className="item   "
      >
        <a
          className="link"
          href="#"
          onClick={[Function]}
        >
          <span
            className="label"
          >
            white1
          </span>
           
          <span
            className="count"
          >
            3
          </span>
        </a>
      </li>
      <li
        className="item   "
      >
        <a
          className="link"
          href="#"
          onClick={[Function]}
        >
          <span
            className="label"
          >
            white2
          </span>
           
          <span
            className="count"
          >
            4
          </span>
        </a>
      </li>
    </ul>
  </li>
  <li
    className="item   "
  >
    <a
      className="link"
      href="#"
      onClick={[Function]}
    >
      <span
        className="label"
      >
        black
      </span>
       
      <span
        className="count"
      >
        20
      </span>
    </a>
  </li>
  <button
    className="showMore false"
    disabled={false}
    onClick={[Function]}
  >
    Show more
  </button>
</ul>
`;
